api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -357,76 +357,100 @@ public Builder withSpecId(int newSpecId) {
       return sourceColumn;
     }
 
-    public Builder identity(String sourceName) {
-      checkAndAddPartitionName(sourceName);
+
+    public Builder identity(String sourceName, String targetName) {
+      checkAndAddPartitionName(targetName);
```

Comment:
When the transform is not `identity`, should we add a check on `checkAndAddPartitionName()` to see if the `targetName` is already a field name in the `schema`? To me, it would be confusing to have a partition name and a schema field be the same when they don't correlate.

Reply:
I think this is a good point. We should not allow a partition field to have the same name as a schema column unless the transform is identity.

Similarly, should we allow overriding the partition name when the transform is identity? I think we should require them to use the same name.

Reply:
done. Took out access to setting custom name on `identity` transform. Added check for non-identity transforms to disallow target names that exist in schema.